### PR TITLE
'Simplify interpolation' should not be suggested in VB for non-literal alignments

### LINKED ISSUE #49712: VB.Net editor fails to use variables in for text width in String interpolation
_This issue has been moved from [a ticket on Developer Community](https://developercommunity2.visualstudio.com/t/VBNet-editor-fails-to-use-variables-in-/1265067)._

---
[severity:It bothers me. A fix would be nice]
I tried a piece of code that works perfectly in C# but not in VB.Net
C# version
! [image.png] (https://aka.ms/dc/image?name=Ba041939db5224da98d2a608d2caa8396637416755140267327_20201122-225832-image.png&tid=a041939db5224da98d2a608d2caa8396637416755140267327)
VB.net version using PadLeft(), you will notice that Visual Studio gives me a tip to simplify the usage of PadLeft()
! [image.png] (https://aka.ms/dc/image?name=B89a9aff59ed14810b0b1a5137a2f5af1637416757412823911_20201122-230219-image.png&tid=89a9aff59ed14810b0b1a5137a2f5af1637416757412823911)
when I follow its advice, I get a compiler error. Somehow the editor forces a "0" before the usage of the constant in the interpolated string

! [image.png] (https://aka.ms/dc/image?name=B0214be849f984f999e4ba253777f93e1637416759153851472_20201122-230513-image.png&tid=0214be849f984f999e4ba253777f93e1637416759153851472)
So it seems the VB.net editor does not comprehend using constants or variables in this place. Please check.

---
#### Feedback Bot on 11/22/2020, 08:24 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Feedback Bot on 11/30/2020, 04:11 PM: 

<p>Thank you for sharing your feedback! Our teams prioritize action on product issues with broad customer impact. See details at: <a target='_blank' href="https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq">https://docs.microsoft.com/en-us/visualstudio/ide/report-a-problem?view=vs-2019#faq</a>. In case you need answers to common questions or need assisted support, be sure to use <a target='_blank' href="https://visualstudio.microsoft.com/vs/support/">https://visualstudio.microsoft.com/vs/support/</a>. We’ll keep you posted on any updates to this feedback.</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by jnm2**
Fixes #49712

💭 Would like to revert https://github.com/dotnet/roslyn/pull/49655/commits/4365779fe54547a53ae7fd10945ac750c63ad09d and bring back abstract helper classes. That way there wouldn't now be three pieces of knowledge duplicated between the analyzers and the fix providers: 1. what conditional syntax is 2. what parenthesized syntax is 3. whether the language supports non-literal alignment components

### CODE REVIEW

File: src/Analyzers/CSharp/Analyzers/SimplifyInterpolation/CSharpHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.SimplifyInterpolation;
+
+namespace Microsoft.CodeAnalysis.CSharp.Analyzers.SimplifyInterpolation
+{
+    internal sealed class CSharpHelpers : AbstractHelpers
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/SimplifyInterpolation/CSharpHelpers.cs:10]: either make these all nested types, or change the names to SimplifyInterpolationHelpers :)  having these vague names is not fun when doing 'nav to' :)

File: src/Analyzers/Core/Analyzers/SimplifyInterpolation/AbstractHelpers.cs [hunk 2]
```diff
@@ -14,28 +14,21 @@
...
-    internal static class Helpers
+    internal abstract class AbstractHelpers
     {
-        private static SyntaxNode GetPreservedInterpolationExpressionSyntax<TConditionalExpressionSyntax, TParenthesizedExpressionSyntax>(
-            IOperation operation)
-            where TConditionalExpressionSyntax : SyntaxNode
-            where TParenthesizedExpressionSyntax : SyntaxNode
+        protected abstract bool PermitNonLiteralAlignmentComponents { get; }
+
+        protected virtual SyntaxNode GetPreservedInterpolationExpressionSyntax(IOperation operation)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/SimplifyInterpolation/AbstractHelpers.cs:21]: make this abstract plz.

File: src/Analyzers/Core/Analyzers/SimplifyInterpolation/AbstractHelpers.cs [hunk 3]
```diff
@@ -160,17 +145,18 @@ private static TextSpan GetSpanWithinLiteralQuotes(IVirtualCharService virtualCh
                             IsSpaceChar(invocation.Arguments[1]))
                         {
                             var alignmentOp = invocation.Arguments[0].Value;
-                            if (alignmentOp != null && alignmentOp.ConstantValue.HasValue)
+                            if (alignmentOp != null && (PermitNonLiteralAlignmentComponents
+                                ? alignmentOp.ConstantValue.HasValue
+                                : alignmentOp.Kind == OperationKind.Literal))
```
Comment (CyrusNajmabadi) [inline src/Analyzers/Core/Analyzers/SimplifyInterpolation/AbstractHelpers.cs:150]: please break this up.  this is too complex for an if-conditional.  thx.

File: src/Analyzers/VisualBasic/Analyzers/SimplifyInterpolation/VisualBasicSimplifyInterpolationDiagnosticAnalyzer.vb [hunk 4]
```diff
@@ -13,11 +13,11 @@ Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
...
     Friend Class VisualBasicSimplifyInterpolationDiagnosticAnalyzer
-        Inherits AbstractSimplifyInterpolationDiagnosticAnalyzer(Of
-            InterpolationSyntax,
-            ExpressionSyntax,
-            TernaryConditionalExpressionSyntax,
-            ParenthesizedExpressionSyntax)
+        Inherits AbstractSimplifyInterpolationDiagnosticAnalyzer(Of InterpolationSyntax, ExpressionSyntax)
+
+        Protected Overrides Function GetHelpers() As AbstractHelpers
+            Return New VisualBasicHelpers
```
Comment (CyrusNajmabadi) [inline src/Analyzers/VisualBasic/Analyzers/SimplifyInterpolation/VisualBasicSimplifyInterpolationDiagnosticAnalyzer.vb:19]:
```suggestion
            Return VisualBasicHelpers.Instance
```

File: src/Analyzers/VisualBasic/CodeFixes/SimplifyInterpolation/VisualBasicSimplifyInterpolationCodeFixProvider.vb [hunk 5]
```diff
@@ -13,14 +13,17 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.SimplifyInterpolation
...
-            TernaryConditionalExpressionSyntax, ParenthesizedExpressionSyntax)
+            InterpolationFormatClauseSyntax, InterpolatedStringExpressionSyntax)
 
         <ImportingConstructor>
         <SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification:="Used in test code: https://github.com/dotnet/roslyn/issues/42814")>
         Public Sub New()
         End Sub
 
+        Protected Overrides Function GetHelpers() As AbstractHelpers
+            Return New VisualBasicHelpers
```
Comment (CyrusNajmabadi) [inline src/Analyzers/VisualBasic/CodeFixes/SimplifyInterpolation/VisualBasicSimplifyInterpolationCodeFixProvider.vb:24]:
```suggestion
            Return VisualBasicHelpers.Instance
```

File: src/Analyzers/Core/Analyzers/SimplifyInterpolation/AbstractHelpers.cs (hunk 3 above)
Comment (jnm2) [reply src/Analyzers/Core/Analyzers/SimplifyInterpolation/AbstractHelpers.cs]: Thanks. Can you give your opinion on the form I replaced it with?

**Comment by jnm2** [comment]
@CyrusNajmabadi Ready for review.

**Comment by jnm2** [comment]
@CyrusNajmabadi Ready for the next step

**Comment by jnm2** [comment]
@sharwell Here's a case where fixing merge conflicts by merging instead of rebasing causes a situation where you really can't see what's happening at all. Do you know where the "merging is better" theory broke down in this case?

**Comment by sharwell** [comment]
@jnm2 if someone had already reviewed the Files Changed for 018c7aa (and marked the files as reviewed in their local view), the re-review of the complete PR (all commits) would only uncheck the items which changed in the final merge commit.

**Comment by jnm2** [comment]
@sharwell Oh, is it something they can benefit from and I can't? I was hoping to be able to have a view of what changed within the PR files in the diff of the merge commit, but it shows me everything, way too much to be useful.

**Comment by sharwell** [comment]
I think the diff is just too big to show the condensed diff. I imagine there is a git command line feature to show the equivalent locally.

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**